physics/processes/inc/pair_production.h

```diff
@@ -25,26 +25,37 @@ __device__ void pair_production::GenerateInteraction(int particle_index, adept::
 {
   track *mytrack = &((*block)[particle_index]);
 
-  float eloss = 0.5f * mytrack->energy;
+  float esecond = 0.5f * mytrack->energy;
 
   // pair production
-  mytrack->energy -= eloss;
+  mytrack->energy -= esecond;
   mytrack->energy_loss           = 0;
-  mytrack->number_of_secondaries = 1;
+  mytrack->number_of_secondaries++;
 
   auto secondary_track = block->NextElement();
   assert(secondary_track != nullptr && "No slot available for secondary track");
-  secondary_track->energy                = eloss;
-  secondary_track->status                = alive;
-  secondary_track->energy_loss           = 0;
-  secondary_track->number_of_secondaries = 0;
-  // Inherit current position and direction.
-  secondary_track->pos           = mytrack->pos;
-  secondary_track->dir           = mytrack->dir;
-  secondary_track->current_state = mytrack->current_state;
-  // Initialize a new PRNG state.
-  secondary_track->rng_state = mytrack->rng_state;
-  secondary_track->rng_state.Skip(1 << 15);
+  if( secondary_track != nullptr){
```

Comment:
This may need a better solution, but for now I think this should really crash (hence the `assert`) because if there's no slot available, we should not silently skip the secondary

Reply:
My problem was that the assert does not exist in non-debug runs.  I got a crash due to this - when I reduced the size of memory allocated to a small value (64K).  

I agree that the simulation must cope with this in a better way.  At least signal that it has failed (exiting in all types of runs, not just when compiled with debugging.)  

We may need a mechanism to 'throttle' the generation of new tracks.  I can imagine several ways.  One would be a kernel that reduces the population of tracks by moving only the lowest-energy tracks, which will typically range out.  This may also happen naturally nearly all the time, so long as these tracks are processed along with the others.  ( Maintaining a healthy number of live tracks is important.) 

But if there is no protection at all when the we are running in "Release" mode, I foresee problems.  Maybe a tally of the number of killed tracks and their total energy, plus an error report if this exceed a non-trivial value would be an ok first step?

Reply:
I would argue that having the code crash is better than silently loosing objects (and their energy).

Maybe we can just exchange the `assert` with something that aborts even in Release mode?

Reply:
I agree with you that this is the better solution.

Reply:
A change in pair_production was made in master that crashes if there's no slot.
Resolved.